File: src/es/greuze/sandbox/prisoners/Prisoner.java
package es.greuze.sandbox.prisoners;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;

public abstract class Prisoner {

    private static final boolean DEBUG = Boolean.TRUE;

    private BufferedReader in;
    private PrintWriter out;

    public void initialize(int port) {
        ServerSocket server = null;
        try {
            server = new ServerSocket(port);
        } catch (IOException e) {
            System.out.println("Could not listen on port " + port);
            System.exit(-1);
        }

        Socket client = null;
        try {
            System.out.println("Waiting for Police...");
            client = server.accept();
        } catch (IOException e) {
            System.out.println("Accept failed: " + port);
            System.exit(-1);
        }

        try {
            in = new BufferedReader(new InputStreamReader(client.getInputStream()));
            out = new PrintWriter(new OutputStreamWriter(client.getOutputStream()), true);
        } catch (IOException e) {
            System.out.println("Read failed");
            System.exit(-1);
        }
    }

    public void startContest() {
        while (mustContinue()) {
            String response = sendAnswer();
            notifyPoliceResponse(response);
        }
    }

    private boolean mustContinue() {
        // Receive data from Police
        log("Waiting instruction...");
        String request = null;
        try {
            request = in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        log("Got '" + request + "'");

        if ("NEXT".equals(request)) {
            return true;
        } else if ("EXIT".equals(request)) {
            return false;
        } else {
            throw new RuntimeException("Unknown operation " + request);
        }
    }

    private String sendAnswer() {
        //Send data back to Police
        String answer = giveAnswer();
        log("Sending '" + answer + "' to Police");
        out.println(answer);

        try {
            return in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void log(String message) {
        if (DEBUG) {
            System.out.println(message);
        }
    }

    // Implement in subclasses
    public abstract String giveAnswer();
    public abstract void notifyPoliceResponse(String response);
}

File: scala/src/Prisoner.scala
package es.greuze.sandbox.prisoners;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.ServerSocket;

abstract class Prisoner {

    def giveAnswer : String
    def notifyPoliceResponse(response: String)

    var in : BufferedReader = null
    var out : PrintWriter = null

    def initialize(port: Int) {
        val serverSocket = new ServerSocket(port)
        println("Waiting for Police on port " + port);
        val socket = serverSocket.accept

        in = new BufferedReader(new InputStreamReader(socket.getInputStream))
        out = new PrintWriter(new OutputStreamWriter(socket.getOutputStream), true)
    }

    def startContest() {
        while (mustContinue()) {
            val answer = giveAnswer
            println("I answer: " + answer)
            out.println(answer)
            val response = in.readLine
            notifyPoliceResponse(response)
        }
    }
    
    def mustContinue() : Boolean = {
        // Receive data from Police
        val request = in.readLine

        if ("NEXT" == request) {
            true;
        } else if ("EXIT" == request) {
            false;
        } else {
            throw new RuntimeException("Unknown operation " + request);
        }
    }
}
